chainer/link_hooks/spectral_normalization.py

```diff
@@ -0,0 +1,264 @@
+    return sigma
+
+
+class SpectralNormalization(link_hook.LinkHook):
+    """Spectral Normalization link hook implementation.
+
+    This hook normalizes a weight using max singular value and this value
+    is computed via power iteration method. Currently, this hook is supposed to
+    be added to :class:`chainer.links.Linear`, :class:`chainer.links.EmbedID`,
+    :class:`chainer.links.Convolution2D`, :class:`chainer.links.ConvolutionND`,
+    :class:`chainer.links.Deconvolution2D`,
+    and :class:`chainer.links.DeconvolutionND`. However, you can use this to
+    other links like RNNs by specifying ``weight_name``.
+    It is highly recommended to add this hook before optimizer setup because
+    this hook add a scaling parameter ``gamma`` if ``use_gamma`` is True.
+    Otherwise, the registered ``gamma`` will not be updated.
+
+    .. math::
+
+       \\bar{\\mathbf{W}} &=& \\dfrac{\\mathbf{W}}{\\sigma(\\mathbf{W})} \\\\
+       \\text{, where} \\ \\sigma(\\mathbf{W}) &:=&
+        \\max_{\\mathbf{h}: \\mathbf{h} \\ne 0}
+       \\dfrac{\\|\\mathbf{W} \\mathbf{h}\\|_2}{\\|\\mathbf{h}\\|_2}
+        = \\max_{\\|\\mathbf{h}\\|_2 \\le 1} \\|\\mathbf{W}\\mathbf{h}\\|_2
+
+    See: T. Miyato et. al., `Spectral Normalization for Generative Adversarial
+    Networks <https://arxiv.org/abs/1802.05957>`_
+
+    Args:
+        n_power_iteration (int): Number of power iteration.
+            The default value is 1.
+        eps (int): Numerical stability in norm calculation.
+            The default value is 1e-12.
+        use_gamma (bool): If ``True``, weight scaling parameter gamma which is
+            initialized by initial weight's max singular value is introduced.
+        factor (float, None): Scaling parameter to divide maximum singular
+            value.  The default value is 1.0.
+        weight_name (str): Link's weight name to apply this hook. The default
+            value is ``'W'``.
+        name (str or None): Name of this hook. The default value is
+            ``'SpectralNormalization'``.
+
+    Attributes:
+        vector_name (str): Name of the approximate first left singular vector
+            registered in the target link.
+            the target link.
+        axis (int): Axis of weight represents the number of output
+            feature maps or output units (``out_channels`` and
+            ``out_size``, respectively).
+
+    .. admonition:: Example
+
+        There are almost the same but 2 ways to apply spectral normalization
+        (SN) hook to links.
+
+        1. Initialize link and SN separately. This makes it easy to handle
+        buffer and parameter of links registered by SN hook.
+
+            >>> l = L.Convolution2D(3, 5, 3)
+            >>> hook = chainer.link_hooks.SpectralNormalization()
+            >>> _ = l.add_hook(hook)
+            >>> # Check the shape of the first left singular vector.
+            >>> getattr(l, hook.vector_name).shape
+            (1, 5)
```

Comment:
The spec has changed during the review:bow:
```suggestion
            (5,)
```